### ISSUE TITLE
Render streamed input incrementally

### ISSUE THREAD

**Issue by david-crespo** [charmbracelet/glow#601]
I wrote a simple [CLI](https://github.com/david-crespo/llm-cli) for talking to LLM APIs and I pipe markdown-formatted output into `glow` for rendering.

```sh
function ai() {
  ~/repos/llm-cli/main.ts "$@" | glow
}
```

I would like to be able to use streaming responses from LLM APIs and write the result to stdout chunk by chunk and have `glow` render it incrementally. But it seems `glow` always reads the full input before rendering:

https://github.com/charmbracelet/glow/blob/2430b0af3fd83bf776875da464a885bdb5f88d38/main.go#L246-L247

I understand why it is that way, I'm sure it's dramatically simpler. It's certainly what I would have done.

### Alternatives

* I could simply not pipe to `glow` when I have a streaming response (or stream raw markdown to stdout and then re-render with `glow` once it's done)
* Use your lovely [`mods`](https://github.com/charmbracelet/mods) CLI instead, which does support streaming but [does not support Claude](https://github.com/charmbracelet/mods/issues/233)

**Comment by david-crespo** [comment]
Update for anyone interested in this: I'm able to solve this outside of `glow` by accumulating the input and clearing and re-rendering the whole thing on each chunk.

```ts
import $ from "jsr:@david/dax@0.41.0"

let inputBuffer = ""

const decoder = new TextDecoder()

for await (const chunk of Deno.stdin.readable) {
  inputBuffer += decoder.decode(chunk)
  // --style auto is there to force it to output styled
  // https://github.com/charmbracelet/glow/blob/2430b0a/main.go#L158
  const output = await $`glow --style auto`.stdinText(inputBuffer).text()
  console.clear()
  console.log(output)
}
```

https://github.com/charmbracelet/glow/assets/3612203/67b888e8-2224-484b-b71b-516690629c4e

**Comment by gwpl** [comment]
Yes! I realised recently that I sometimes pipe results of LLMs thought `glow` to have more coloring and why it can't be streaming support enabled by default! It would be great :).

**Comment by kaiwen-wang** [comment]
I using Claude to re-write your solution for fish and `llm` by Simon Willison, but I get a re-render flashing in glow. The `llm-cli` you made is cool though it seems to do thinking rather than streaming.

**Comment by nbbaier** [comment]
> I using Claude to re-write your solution for fish and `llm` by Simon Willison, but I get a re-render flashing in glow.

I'm interested in streaming support for the same reason!

**Comment by n-WN** [comment]
This is my little toy, feel free to try [aimd](https://github.com/n-WN/aimd)

> > I using Claude to re-write your solution for fish and `llm` by Simon Willison, but I get a re-render flashing in glow.
> 
> I'm interested in streaming support for the same reason!

**Comment by anthonyrisinger** [comment]
@david-crespo (and others) See https://github.com/charmbracelet/glow/pull/823 and LMK if it works for you; I'm hoping we can rally the various streaming-related issues around this one, because it looks like at least 2-3+ issues will close if this patch lands (possibly more because it adds some process-level quality-of-life things like signal handling) ☺️

**Comment by cjccjj** [comment]
I made this [https://github.com/cjccjj/mdflow ](https://github.com/cjccjj/mdflow ) 
It renders Markdown in real time streaming, no buffering, no rewrite, well tested againt commomMark and GFM, and super fast. 
I hope this is helpful.

**Comment by llimllib** [comment]
@cjccjj i made a similar program: https://github.com/llimllib/mdriver yours looks nice too

**Comment by cjccjj** [comment]
> [@cjccjj](https://github.com/cjccjj) i made a similar program: https://github.com/llimllib/mdriver yours looks nice too

@llimllib Thanks! I just tried `mdriver`. The features are cool. However, my [`mdflow`](https://github.com/cjccjj/mdflow) has a few advantages:

- `mdflow` streams everything, including code blocks, quotes, and more.
- It is tested against the CommonMark spec, which is a big and difficult task, thanks to MD4C, the proven parser on which it is based.
- Its tables support word wrapping.
- The binary is only about 300 KB, uses at most around 2.5 MB of memory regardless of input size, and is very fast. This is again largely thanks to MD4C. (Check the README in the `mdflow` repo: `glow` uses almost 1 GB of RAM for a 10 MB input. 🤣)
